### DISCUSSION TITLE
Onboard LQR Attitude Controller

### DISCUSSION THREAD

**Discussion by FrancisGoncalves** [bitcraze/discussions#292] [Q&A] [answered]
Hi, I am trying to implement an onboard LQR attitude controller. I ran simulations in Simulink with the Crazyflie Parameters and the controller does a good job stabilizing the crazyflie.
First, I just wanted to verify that it is working so I have all my controller inside the stabilizer.c file. Then, because it was unstable, I decided to try to implement your PID Controller in the stabilizer.c file and it is also unstable. I am not sure if I have something wrong with my code, but I don't think so. Is there a problem (or delay) with doing the controller calculations in the stabilizer.c file? Should I try to implement the controller as you have implemented yours?

Thank you very much,
Francisco

**Comment by krichardsson** [comment, marked as answer]
Hi!

It should be fine to implement in the stabilizer.c but to me it would make more sense to add a new controller as you can switch back and forth between controllers and so on. 

There is no magic here, the call to [`controller()`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L275) is simply calling the controller function in what ever controller that is currently active, [`controllerPid()`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_pid.c#L56) for instance if the PID controller is used. This means that is no difference if your code is implemented in stabilizer.c or a controller (in terms of delays or similar).

If you want to register your own controller just copy an existing controller and change it to your own implementation, then add it to [the list of controllers](https://github.com/bitcraze/crazyflie-firmware/blob/b370e20e50106dc10837c97e28404ff5d3efc5ea/src/modules/src/controller.c#L24-L29). After this it should be possible to switch to your controller by setting the "stabilizer.controller" parameter.

It is hard to say why your implementation is unstable, but maybe it is because you are doing too much work and preventing the stabilizer loop to run at full speed? If this is the case you should get a warning "WARNING: stabilizer loop rate is off" in the console log.

The stabilizer loop runs at 1000Hz and I guess it is likely that you might not want all your controller code to execute that often. You can see [an example in the PID controller](https://github.com/bitcraze/crazyflie-firmware/blob/b370e20e50106dc10837c97e28404ff5d3efc5ea/src/modules/src/controller_pid.c#L61) of the use of the `RATE_DO_EXECUTE` macro that you can use if you want some part of your code to run at a lower rate.

**Reply by FrancisGoncalves** [reply]
Thank you for replying, I did not see any 'stabilizer loop rate is off' warning, but I will follow your suggestions and let you know how it worked. Do you think the issue might be because your gains are tuned for a rate of 500Hz and I was running it at 1000Hz?

**Reply by knmcguire** [reply]
Hi! Let me jump in here. 

Theoretically, it should not make a difference if gains are tuned for 500 Hz or 1000 hz as the controller should be controlling on the state. However, implementation is something that needs to be checked. The difference in time is supposed to handle the scaling of PID gains but if that is wrong compared to reality then you will probably have instability. 

https://github.com/bitcraze/crazyflie-firmware/blob/b370e20e50106dc10837c97e28404ff5d3efc5ea/src/modules/src/position_controller_pid.c#L74

Perhaps that is what you should double check if the difference in time used is the difference in reality.

**Reply by FrancisGoncalves** [reply]
Thank you! I was able to solve the problem just by implementing the controller in functions as you have yours.
